pkgs/shells/nushell/plugins/default.nix

```
@@ -22,25 +24,154 @@ lib.makeScope newScope (
+                        inpkg && /^version = "/ { vr=$0; sub(/^version = "/,"",vr); sub(/"$/,"",vr) }
+                        inpkg && NF==0          { if(nm==n){print vr; exit 0} inpkg=0 }
+                        END { if(nm==n) print vr }
+                      ' "$root_dir/Cargo.lock"
+                    )" || return 1
+                    [ -z "$locked" ] && return 1
+
+                    is_semvers_compatible "$locked" "$want"
+                    case $? in
+                      0) return 0 ;;
+                      1) return 1 ;;
+                      *) return 0 ;;
+                    esac
+                  }
+
+                  patch_crate_version() { # crate, version
+                    local crate="$1" ver="$2"
+
+                    find "$root_dir" -name Cargo.toml -print0 |
+                    while IFS= read -r -d ''' toml; do
+                      sed -i -E "s|^([[:space:]]*$crate[[:space:]]*=[[:space:]]*\")[^\"]*(\".*)|\1$ver\2|" "$toml" || true
+                      sed -i -E "s|^([[:space:]]*$crate[[:space:]]*=[[:space:]]*\{[^}]*version[[:space:]]*=[[:space:]]*\")[^\"]*(\"[^}]*\})|\1$ver\2|" "$toml" || true
+                    done
+                  }
+                '';
               in
-              runCommand "test-load-${final.pname}" { } ''
-                touch $out
-                ${nu} -n -c "plugin add --plugin-config $out ${plugin}"
-                ${nu} -n -c "plugin use --plugin-config $out ${plugin}"
-              '';
-          }
-        );
+              {
+
+                cargoDeps = rustPlatform.fetchCargoVendor ({
+                  inherit (prev) src;
+                  name = "${prev.pname}-${prev.version}";
+                  hash = prev.cargoHash;
+                  nativeBuildInputs = [ cargo ];
+
+                  postPatch = ''
+                    ${patchCrateVersionHelper}
+
+                    if is_compatible nu-plugin "${nushellVersion}"; then
+                      echo "nu-plugin is compatible, no patching needed"
+                    else
+                      patch_crate_version nu-plugin "${nushellVersion}"
+                      patch_crate_version nu-protocol "${nushellVersion}"
+                      cargo update -w -p nu-plugin
+                      cargo update -w -p nu-protocol
+                      cargo generate-lockfile
+                    fi
+                  '';
+                });
+
+                postPatch = ''
+                  ${patchCrateVersionHelper}
+
+                  if is_compatible nu-plugin "${nushellVersion}"; then
+                    echo "nu-plugin is compatible, no patching needed"
+                  else
+                    patch_crate_version nu-plugin "${nushellVersion}"
+                    patch_crate_version nu-protocol "${nushellVersion}"
+                    cp ${final.cargoDeps}/Cargo.lock ./Cargo.lock
+                  fi
+                '';
```

This seems unnecessarily duplicate. Can't you just extract it into a single variable (so it does the same twice), and do this (would probably work best if you also used `cargo-edit` like I suggested above):
```suggestion
                  inherit postPatch;
                });

                inherit postPatch;
```